allennlp/modules/similarity_functions/multiheaded.py

```diff
@@ -0,0 +1,114 @@
+        into multiple heads.  This number must be divisible evenly by ``num_heads``.  If not given,
+        we default to ``tensor_1_dim``.
+    tensor_2_dim : ``int``, optional
+        The dimension of the second tensor described above.  This is ``tensor.size()[-1]`` - the
+        length of the vector `before` the multi-headed projection.  We need this so we can build
+        the weight matrix correctly.  If not given, we default to ``tensor_1_dim``.
+    tensor_2_projected_dim : ``int``, optional
+        The dimension of the second tensor `after` the multi-headed projection, `before` we split
+        into multiple heads.  This number must be divisible evenly by ``num_heads``.  If not given,
+        we default to ``tensor_2_dim``.
+    internal_similarity : ``SimilarityFunction``, optional
+        The ``SimilarityFunction`` to call on the projected, multi-headed tensors.  The default is
+        to use a dot product.
+    """
+    def __init__(self,
+                 num_heads: int,
+                 tensor_1_dim: int,
+                 tensor_1_projected_dim: int = None,
+                 tensor_2_dim: int = None,
+                 tensor_2_projected_dim: int = None,
+                 internal_similarity: SimilarityFunction = DotProductSimilarity()) -> None:
+        super(MultiHeadedSimilarity, self).__init__()
+        self._num_heads = num_heads
+        self._internal_similarity = internal_similarity
+        tensor_1_projected_dim = tensor_1_projected_dim or tensor_1_dim
+        tensor_2_dim = tensor_2_dim or tensor_1_dim
+        tensor_2_projected_dim = tensor_2_projected_dim or tensor_2_dim
+        if tensor_1_projected_dim % num_heads != 0:
+            raise ConfigurationError("Projected dimension not divisible by number of heads: %d, %d"
+                                     % (tensor_1_projected_dim, num_heads))
+        if tensor_2_projected_dim % num_heads != 0:
+            raise ConfigurationError("Projected dimension not divisible by number of heads: %d, %d"
+                                     % (tensor_2_projected_dim, num_heads))
+        self._tensor_1_projection = Parameter(torch.Tensor(tensor_1_dim, tensor_1_projected_dim))
+        self._tensor_2_projection = Parameter(torch.Tensor(tensor_2_dim, tensor_2_projected_dim))
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        torch.nn.init.xavier_uniform(self._tensor_1_projection)
+        torch.nn.init.xavier_uniform(self._tensor_2_projection)
+
+    @overrides
+    def forward(self, tensor_1: torch.Tensor, tensor_2: torch.Tensor) -> torch.Tensor:
+        projected_tensor_1 = torch.matmul(tensor_1, self._tensor_1_projection)
+        projected_tensor_2 = torch.matmul(tensor_2, self._tensor_2_projection)
+
+        # split_tensor_{1,2} are both tuples of length num_heads, where each element in the tuple
+        # is a tensor of shape (..., projected_dim / num_heads).
+        split_size = projected_tensor_1.size(-1) // self._num_heads
```

Comment:
You don't need this split and stack - you can just do 
```
all_but_final_dim = projected_tensor_x.size()[:-1]
projected_tensor_x.resize(*all_but_final_dim, self._num_heads, -1)
```

Reply:
Right, good point, thanks.